scripts/terraform_schema.py

```diff
@@ -0,0 +1,74 @@
+"""
+Save Terraform version and schema information to the file specified
+in AZUL_TRACKED_SCHEMA_PATH or verify that the Terraform information
+in that file is up-to-date.
+"""
+import argparse
+import json
+import logging
+import os.path
+import subprocess
+
+from azul import (
+    config,
+)
+
+from azul.logging import (
+    configure_script_logging,
+)
+
+log = logging.getLogger(__name__)
+
+
+def terraform(*args: str) -> str:
+    terraform_dir = os.path.join(config.project_root, 'terraform')
+    cmd = subprocess.run(['terraform', *args],
+                         cwd=terraform_dir,
+                         check=True,
+                         stdout=subprocess.PIPE,
+                         text=True,
+                         shell=False)
+    return cmd.stdout
+
+
+def terraform_version() -> str:
+    # `terraform -version` prints a warning if you are not running the latest
+    # release of Terraform; we discard it, otherwise, we would need to update
+    # the tracked schema every time a new version of Terraform is released
+    versions, footer = terraform('-version').split('\n\n')
+    return versions
+
+
+def generate_schema_json(path: str) -> None:
+    schema = terraform('providers', 'schema', '-json')
+    tracked = {
+        'versions': terraform_version(),
+        'schema': json.loads(schema)
+    }
+    with open(path, 'w') as f:
+        json.dump(tracked, f, indent=4)
+
+
+def check_schema_json(path: str) -> None:
+    with open(path, 'r') as f:
+        versions_tracked = json.load(f)['versions']
+    if terraform_version() == versions_tracked:
+        return
+    else:
+        raise RuntimeError('Tracked Terraform schema is out of date. Run '
+                           f'`make -C terraform schema` and commit {path}')
+
+
+if __name__ == '__main__':
+    configure_script_logging()
+    commands = {
+        'generate': generate_schema_json,
+        'verify': check_schema_json
+    }
+    # https://youtrack.jetbrains.com/issue/PY-41806
+    # noinspection PyTypeChecker
+    parser = argparse.ArgumentParser(description=__doc__,
+                                     formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+    parser.add_argument('command', choices=commands)
+    arguments = parser.parse_args()
+    commands[arguments.command](config.tracked_schema_path)
```

Comment:
… since this is the only call site and I don't think anyone would want to call the function with a different value.